### PR TITLE
Buffered writes while awaiting for other appenders to release the header

### PR DESCRIPTION
**Pull request by JerryShea**

### CODE REVIEW

**Comment by JerryShea** [comment]
Some initial perf results from net.openhft.chronicle.queue.ContendedWriterTest
Before
```
oneThread
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,560 / 5,900  14,420 - 14,420
twoThreadsWritingLargeMessagesAtSameSlowRate
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  3,470 / 5,370  5,370 - 5,370
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  3,600 / 5,900  5,900 - 5,900
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlow
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,160 / 5,900  9,180 - 9,180
ORIGINAL 50/90 99/99.9 99.99 - worst was 418 / 1,740  9,180 / 25,690  25,690 - 25,690
twoThreadsWritingLargeMessagesOneFastOneSlow
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,080 / 2,160  4,850 / 13,890  17,300 - 17,300
ORIGINAL 50/90 99/99.9 99.99 - worst was 1,540 / 4,330  21,500 / 180,360  180,360 - 180,360
```
Enable deferred header writes
```
oneThread
DEFFERED 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,690 / 8,650  11,800 - 11,800
twoThreadsWritingLargeMessagesAtSameSlowRate
DEFFERED 50/90 99/99.9 99.99 - worst was 1,150 / 1,470  5,110 / 11,800  11,800 - 11,800
DEFFERED 50/90 99/99.9 99.99 - worst was 1,150 / 1,470  5,370 / 11,800  11,800 - 11,800
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlow
DEFFERED 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  1,610 / 6,420  10,750 - 10,750
DEFFERED 50/90 99/99.9 99.99 - worst was 451 / 639  4,590 / 10,750  10,750 - 10,750
twoThreadsWritingLargeMessagesOneFastOneSlow
DEFFERED 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  3,340 / 5,640  8,260 - 8,260
DEFFERED 50/90 99/99.9 99.99 - worst was 1,540 / 5,900  31,980 / 77,590  77,590 - 77,590
```
The numbers jump around a bit. I have seen deferred writes win out but last few runs they didn't. The test scenario needs a bit more work I think.

**Comment by peter-lawrey** [comment]
If the number jump around i suspect they are not run long enough. Also i expect this wiill make more difference with more threads. Eg 6 which i think is the use case.

**Comment by JerryShea** [comment]
Done some improvements to config and re-tested:
Single thread numbers are pretty much the same as each other - in this case deferred is faster, but not every time
```
oneThread
50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,160 / 5,640  12,320 - 12,320
oneThreadDeferred
50/90 99/99.9 99.99 - worst was 1,080 / 1,150  1,610 / 4,330  8,260 - 8,260
---
twoThreadsWritingLargeMessagesAtSameSlowRate
thread1 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,290 / 5,640  5,640 - 5,640
thread2 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  2,690 / 5,640  5,640 - 5,640
twoThreadsWritingLargeMessagesAtSameSlowRateBothDeferred
thread1 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 1,210  3,210 / 36,700  36,700 - 36,700
thread2 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 1,210  3,080 / 4,330  4,330 - 4,330
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlow
thread1 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  1,610 / 4,590  5,370 - 5,370
thread2 progress=false writePause=0 between=5: 50/90 99/99.9 99.99 - worst was 418 / 1,210  10,220 / 26,740  26,740 - 26,740
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlowAndDeferred
thread1 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  1,340 / 4,330  8,650 - 8,650
thread2 progress=true writePause=0 between=5: 50/90 99/99.9 99.99 - worst was 434 / 639  4,590 / 10,220  10,220 - 10,220
twoThreadsWritingLargeMessagesOneFastOneSlow
thread1 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,150 / 2,160  2,420 / 5,370  13,370 - 13,370
thread2 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,540 / 1,740  4,850 / 38,800  38,800 - 38,800
twoThreadsWritingLargeMessagesOneFastOneSlowAndDeferred
thread1 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,080 / 1,150  1,800 / 5,370  8,650 - 8,650
thread2 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,470 / 2,690  13,890 / 81,790  81,790 - 81,790

```

File: src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java [hunk 1]
```diff
@@ -862,6 +858,87 @@ public String toString() {
...
+                    assert wire != null;
+                    assert wire != context.wire;
+                    for (int i = 0; i < 128; i++) {
+                        try {
+                            // TODO: we should be able to write and update the header in one go
+                            long pos = store.writeHeader(wire, Wires.UNKNOWN_LENGTH, safeLength, timeoutMS());
+                            position(pos);
+                            long length = context.wire.bytes().copyTo(wire.bytes());
+                            wire.bytes().writePosition(length + wire.bytes().writePosition());
+                            context.wire = wire;
```
Comment (epickrram) [inline src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java]: Should you reset the context wire in a finally block? In the event of an exception, you'll be left using the instance returned from acquireBufferWire()?

**Comment by JerryShea** [comment]
Ran for longer and added a couple of 6 thread tests. We definitely saw problems with 6 components writing, but another one that was problematic was 2 components, one writing fast and one writing slow (see twoThreadsWritingLargeMessagesFastAndSmallMessagesSlow). The sixThreads tests show this new feature as working *very* effectively.
```
net.openhft.chronicle.queue.ContendedWriterTest
oneThread
thread0 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,150 / 1,470  4,330 / 7,210  10,220 - 13,890
oneThreadDeferred
thread0 progress=true writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,150 / 1,610  5,110 / 8,260  10,750 - 12,320
sixThreads
thread0 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 13,370 / 45,090  119,540 / 155,190  155,190 - 155,190
thread1 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 10,750 / 45,090  85,980 / 171,970  171,970 - 171,970
thread2 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 12,850 / 49,280  102,760 / 171,970  171,970 - 171,970
thread3 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 10,220 / 45,090  90,180 / 138,410  138,410 - 138,410
thread4 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 11,270 / 47,190  111,150 / 188,740  188,740 - 188,740
thread5 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 10,220 / 47,190  98,570 / 146,800  146,800 - 146,800
sixThreadsDeferred
thread0 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 2,820  13,890 / 34,600  127,930 - 127,930
thread1 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 2,820  14,940 / 34,600  111,150 - 111,150
thread2 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 3,080  15,990 / 29,880  115,340 - 115,340
thread3 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 3,080  14,940 / 28,840  106,950 - 106,950
thread4 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 2,820  14,940 / 28,840  90,180 - 90,180
thread5 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,150 / 2,690  14,940 / 30,930  38,800 - 38,800
twoThreadsWritingLargeMessagesAtSameSlowRate
thread0 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,210 / 3,600  9,700 / 17,300  19,400 - 19,400
thread1 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,210 / 3,470  9,700 / 21,500  29,880 - 29,880
twoThreadsWritingLargeMessagesAtSameSlowRateBothDeferred
thread0 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,210 / 4,590  9,700 / 15,990  16,520 - 16,520
thread1 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 1,210 / 4,590  9,700 / 22,540  26,740 - 26,740
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlow
thread0 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,150 / 1,740  5,110 / 8,000  10,750 - 13,370
thread1 progress=false writePause=0 between=5: 50/90 99/99.9 99.99 - worst was 836 / 8,260  111,150 / 293,600  310,380 - 310,380
twoThreadsWritingLargeMessagesFastAndSmallMessagesSlowAndDeferred
thread0 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,080 / 1,800  4,850 / 8,260  11,800 - 13,370
thread1 progress=true writePause=0 between=5: 50/90 99/99.9 99.99 - worst was 803 / 6,420  239,080 / 1,442,840  1,442,840 - 1,442,840
twoThreadsWritingLargeMessagesOneFastOneSlow
thread0 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,150 / 2,690  8,650 / 18,350  26,740 - 26,740
thread1 progress=false writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 2,290 / 14,420  127,930 / 264,240  310,380 - 310,380
twoThreadsWritingLargeMessagesOneFastOneSlowAndDeferred
thread0 progress=false writePause=1 between=0: 50/90 99/99.9 99.99 - worst was 1,080 / 1,540  5,110 / 8,000  10,220 - 11,270
thread1 progress=true writePause=1 between=5: 50/90 99/99.9 99.99 - worst was 2,290 / 17,300  146,800 / 327,160  327,160 - 327,160
```

File: src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java (hunk 1 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java]: This method is called when we close the DocumentContext, typically in at the end of a try-with-resources, so it _shouldn't_ be used any more. When we open the DocumentContext, we set context.wire. Does that address your concerns?

**Comment by JerryShea** [comment]
Have made this configurable with net.openhft.chronicle.queue.ChronicleQueueBuilder#progressOnContention()
```
Setting this true will likely improve throughput if
    you have multiple appenders, and
    the appenders are holding the writingContext open for a long time (e.g. large objects that are slow to serialise)
```

File: src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java (hunk 1 above)
Comment (epickrram) [reply src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java]: No, that's fine, if it's the terminal state for that instance, then there's no need for any change.

**Pull request merged**